### ISSUE TITLE
Dynamic Threshold

### ISSUE THREAD

**Issue by KaiSUN1** [fangwei123456/spikingjelly#564]
I want to get adaptive threshold, but why the backpropagation threshold has no gradient, who can help me?

class ALIFNode(neuron.BaseNode):
    def __init__(self, tau: torch.Tensor, v_threshold: float = 1.0, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.v_threshold = torch.nn.Parameter(torch.tensor(v_threshold, dtype=torch.float32, requires_grad=True))
        if not isinstance(tau, torch.Tensor):
            tau = torch.tensor(tau, dtype=torch.float32)
        self.register_buffer('tau', tau)
        self.v_reset = torch.tensor(0.0)  # Assuming a reset value, you might need to adjust this

    def neuronal_charge(self, x: torch.Tensor):
        self.v = self.v + (x - (self.v - self.v_reset)) / self.tau

    def neuronal_fire(self):
        # print(self.v_threshold)
        return self.surrogate_function(self.v - self.v_threshold)

**Comment by frostylight** [comment]
It seems that it works for me, you should provide the whole codes that can reproduce the bug.

```python
class ALIFNode(neuron.BaseNode):
	def __init__(self, tau: float | torch.Tensor = 2., v_threshold: float = 1., *args, **kwargs) -> None:
		super().__init__(*args, **kwargs)

		self.v_threshold = nn.Parameter(torch.tensor(v_threshold, dtype=torch.float, requires_grad=True))
		tau = torch.tensor(tau, dtype=torch.float)
		self.register_buffer("tau", tau)

	def neuronal_charge(self, x: torch.Tensor):
		self.v = self.v + (x - (self.v - self.v_reset)) / self.tau

	def neuronal_fire(self):
		return self.surrogate_function(self.v - self.v_threshold)

torch.manual_seed(0)
an = ALIFNode()
print(an.v_threshold.grad)
result = an(torch.tensor(1))
result.backward()
print(an.v_threshold.grad)
```

the output is
```python
None
tensor(-0.4200)
```

**Comment by CheKaiWei** [comment]
It is correct and should work. You can print the parameters in your optimizer to check if ``v_threshold`` is included.
> 		self.v_threshold = nn.Parameter(torch.tensor(v_threshold, dtype=torch.float, requires_grad=True))

**Comment by AllenYolk** [comment]
`v_threshold` 在 `BaseNode.__init__` 中是 `float`，没有 `requires_grad=True`，因此不会有梯度。

实现可学习阈值有几种方法：

**方法 1**：用 `nn.Parameter` 覆盖 `v_threshold`

```python
class ALIFNode(neuron.BaseNode):
    def __init__(self, tau: float = 2., v_threshold: float = 1., *args, **kwargs):
        super().__init__(v_threshold, *args, **kwargs)
        self.v_threshold = nn.Parameter(
            torch.tensor(v_threshold, dtype=torch.float)
        )
        self.register_buffer("tau", torch.tensor(tau, dtype=torch.float))

    def neuronal_charge(self, x: torch.Tensor):
        self.v = self.v + (x - (self.v - self.v_reset)) / self.tau

    def neuronal_fire(self):
        return self.surrogate_function(self.v - self.v_threshold)
```

注意覆盖后，`BaseNode` 的 `soft_reset`（`v_reset is None` 时减去 `v_threshold`）依然能正确读 `self.v_threshold`。

**方法 2**：参考现成的 `ParametricLIFNode`（位于 `spikingjelly.activation_based.neuron`）。它把 `1/τ = Sigmoid(w)` 作为可学习参数，可以视为"阈值相关参数化"的另一种实现：

```python
from spikingjelly.activation_based import neuron
node = neuron.ParametricLIFNode(init_tau=2.0)
```

如果还需要在 LIF 中同时学习 `τ` 和 `v_threshold`，也可以继承 `ParametricLIFNode` 并把 `v_threshold` 改成 `nn.Parameter`。

**Comment by AllenYolk** [comment]
`v_threshold` is a plain `float` in `BaseNode.__init__` (no `requires_grad=True`), so it has no gradient by default.

Two ways to get a learnable threshold:

**Option 1**: override `v_threshold` with `nn.Parameter`:

```python
class ALIFNode(neuron.BaseNode):
    def __init__(self, tau: float = 2., v_threshold: float = 1., *args, **kwargs):
        super().__init__(v_threshold, *args, **kwargs)
        self.v_threshold = nn.Parameter(
            torch.tensor(v_threshold, dtype=torch.float)
        )
        self.register_buffer("tau", torch.tensor(tau, dtype=torch.float))

    def neuronal_charge(self, x: torch.Tensor):
        self.v = self.v + (x - (self.v - self.v_reset)) / self.tau

    def neuronal_fire(self):
        return self.surrogate_function(self.v - self.v_threshold)
```

Note that after the override, `BaseNode`'s soft-reset path (`v_reset is None` → subtract `v_threshold`) still reads `self.v_threshold` correctly.

**Option 2**: use the built-in `ParametricLIFNode` (in `spikingjelly.activation_based.neuron`), which parameterizes `1/τ = Sigmoid(w)` as a learnable weight — that's another common "threshold-related parameterization":

```python
from spikingjelly.activation_based import neuron
node = neuron.ParametricLIFNode(init_tau=2.0)
```

If you want both learnable `τ` and `v_threshold` in one LIF, you can also inherit from `ParametricLIFNode` and replace `v_threshold` with `nn.Parameter`.
